File: Provenance/Controller/PVPSXControllerViewController.m
//
//  PVPSXControllerViewController.m
//  Provenance
//
//  Created by shruglins on 26/8/16.
//  Copyright © 2016 James Addyman. All rights reserved.
//

#import "PVPSXControllerViewController.h"
#import <PVMednafen/MednafenGameCore.h>

@interface PVPSXControllerViewController ()

@end

@implementation PVPSXControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue; // skip over the PVButtonGroupOverlayView
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"✖"])
        {
            [button setTag:PVPSXButtonCross];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"●"])
        {
            [button setTag:PVPSXButtonCircle];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"◼"])
        {
            [button setTag:PVPSXButtonSquare];
        }
        else if ([[[button titleLabel] text] isEqualToString:@"▲"])
        {
            [button setTag:PVPSXButtonTriangle];
        }
    }
    
    [self.leftShoulderButton setTag:PVPSXButtonL1];
    [self.rightShoulderButton setTag:PVPSXButtonR1];

    [self.selectButton setTag:PVPSXButtonSelect];
    [self.startButton setTag:PVPSXButtonStart];
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    MednafenGameCore *psxCore = (MednafenGameCore *)self.emulatorCore;
    
    [psxCore didReleasePSXButton:PVPSXButtonUp forPlayer:0];
    [psxCore didReleasePSXButton:PVPSXButtonDown forPlayer:0];
    [psxCore didReleasePSXButton:PVPSXButtonLeft forPlayer:0];
    [psxCore didReleasePSXButton:PVPSXButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [psxCore didPushPSXButton:PVPSXButtonUp forPlayer:0];
            [psxCore didPushPSXButton:PVPSXButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [psxCore didPushPSXButton:PVPSXButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [psxCore didPushPSXButton:PVPSXButtonUp forPlayer:0];
            [psxCore didPushPSXButton:PVPSXButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [psxCore didPushPSXButton:PVPSXButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [psxCore didPushPSXButton:PVPSXButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [psxCore didPushPSXButton:PVPSXButtonDown forPlayer:0];
            [psxCore didPushPSXButton:PVPSXButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [psxCore didPushPSXButton:PVPSXButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [psxCore didPushPSXButton:PVPSXButtonDown forPlayer:0];
            [psxCore didPushPSXButton:PVPSXButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
    [self vibrate];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    MednafenGameCore *psxCore = (MednafenGameCore *)self.emulatorCore;
    
    [psxCore didReleasePSXButton:PVPSXButtonUp forPlayer:0];
    [psxCore didReleasePSXButton:PVPSXButtonDown forPlayer:0];
    [psxCore didReleasePSXButton:PVPSXButtonLeft forPlayer:0];
    [psxCore didReleasePSXButton:PVPSXButtonRight forPlayer:0];
}

- (void)buttonPressed:(JSButton *)button
{
    MednafenGameCore *psxCore = (MednafenGameCore *)self.emulatorCore;
    [psxCore didPushPSXButton:[button tag] forPlayer:0];
    
    [self vibrate];
}

- (void)buttonReleased:(JSButton *)button
{
    MednafenGameCore *psxCore = (MednafenGameCore *)self.emulatorCore;
    [psxCore didReleasePSXButton:[button tag] forPlayer:0];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *psxCore = (MednafenGameCore *)self.emulatorCore;
    [psxCore didPushPSXButton:PVPSXButtonStart forPlayer:player];
	psxCore.isStartPressed = YES;
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    MednafenGameCore *psxCore = (MednafenGameCore *)self.emulatorCore;
    [psxCore didReleasePSXButton:PVPSXButtonStart forPlayer:player];
	psxCore.isStartPressed = NO;
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *psxCore = (MednafenGameCore *)self.emulatorCore;
    [psxCore didPushPSXButton:PVPSXButtonSelect forPlayer:player];
	psxCore.isSelectPressed = YES;
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    MednafenGameCore *psxCore = (MednafenGameCore *)self.emulatorCore;
    [psxCore didReleasePSXButton:PVPSXButtonSelect forPlayer:player];
	psxCore.isSelectPressed = NO;
}

@end

File: Provenance/Controller/PVPSXControllerViewController.swift
//  Converted to Swift 4 by Swiftify v4.1.6640 - https://objectivec2swift.com/
//
//  PVPSXControllerViewController.swift
//  Provenance
//
//  Created by shruglins on 26/8/16.
//  Copyright © 2016 James Addyman. All rights reserved.
//

import PVSupport

fileprivate extension JSButton {
    var buttonTag: PVPSXButton {
        get {
            return PVPSXButton(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVPSXControllerViewController: PVControllerViewController<PVPSXSystemResponderClient> {

    override func layoutViews() {
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let text = button.titleLabel.text else {
                return
            }
            if text == "✖" || text == "✕" {
                button.buttonTag = .cross
            } else if text == "●" || text == "○" {
                button.buttonTag = .circle
            } else if text == "◼" || text == "□" {
                button.buttonTag = .square
            } else if text == "▲" || text == "▵" {
                button.buttonTag = .triangle
            }
        }

        leftShoulderButton?.buttonTag = .l1
        rightShoulderButton?.buttonTag = .r1
        leftShoulderButton2?.buttonTag = .l2
        rightShoulderButton2?.buttonTag = .r2
        selectButton?.buttonTag = .select
        startButton?.buttonTag = .start
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
        switch direction {
            case .upLeft:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .up:
                emulatorCore.didPush(.up, forPlayer: 0)
            case .upRight:
                emulatorCore.didPush(.up, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            case .left:
                emulatorCore.didPush(.left, forPlayer: 0)
            case .right:
                emulatorCore.didPush(.right, forPlayer: 0)
            case .downLeft:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.left, forPlayer: 0)
            case .down:
                emulatorCore.didPush(.down, forPlayer: 0)
            case .downRight:
                emulatorCore.didPush(.down, forPlayer: 0)
                emulatorCore.didPush(.right, forPlayer: 0)
            default:
                break
        }
        vibrate()
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        emulatorCore.didRelease(.up, forPlayer: 0)
        emulatorCore.didRelease(.down, forPlayer: 0)
        emulatorCore.didRelease(.left, forPlayer: 0)
        emulatorCore.didRelease(.right, forPlayer: 0)
    }

    override func buttonPressed(_ button: JSButton) {
        emulatorCore.didPush(button.buttonTag, forPlayer: 0)
        super.buttonPressed(button)
    }

    override func buttonReleased(_ button: JSButton) {
        emulatorCore.didRelease(button.buttonTag, forPlayer: 0)
    }

    override func pressStart(forPlayer player: Int) {
        emulatorCore.didPush(.start, forPlayer: player)
        super.pressStart(forPlayer: player)
    }

    override func releaseStart(forPlayer player: Int) {
        emulatorCore.didRelease(.start, forPlayer: player)
    }

    override func pressSelect(forPlayer player: Int) {
        emulatorCore.didPush(.select, forPlayer: player)
        super.pressSelect(forPlayer: player)
    }

    override func releaseSelect(forPlayer player: Int) {
        emulatorCore.didRelease(.select, forPlayer: player)
    }
}
